_includes/manuals/2.4/3.6_upgrade.md

```diff
@@ -31,113 +52,136 @@ For more information about how to backup your instance head over to
 
-In order to catch all configuration changes and manage them properly you should install and run
-rpmconf from the EPEL repository along with vim-enhanced (for vimdiff):
-
-    rpmconf -a --frontend=vimdiff
-
-You can skip to Step 3.
+{% highlight bash %}
+yum erase rh-ruby22\* rh-ruby24\* rh-ror42\* tfm-ror52\*
+{% endhighlight %}
+</div>
 
-##### Step 2 (B) - Debian or Ubuntu package (deb) and installer setups
+<div class="upgrade_os upgrade_os_el8">
+To upgrade an existing Foreman installation, first update with the
+appropriate foreman-release package:
 
-Upgrading from the last release to {{page.version}} has been tested on both
-Debian and Ubuntu. Updating the packages will upgrade the application and
-automatically migrate the database.
+{% highlight bash %}
+dnf upgrade https://yum.theforeman.org/releases/{{page.version}}/el8/x86_64/foreman-release.rpm
+{% endhighlight %}
 
-First edit `/etc/apt/sources.list.d/foreman.list` and change the release
-number from the previous release or "stable" to `{{ page.version }}`:
+Clean up the yum metadata cache:
 
-    deb http://deb.theforeman.org/ buster {{ page.version }}
-    deb http://deb.theforeman.org/ plugins {{ page.version }}
+{% highlight bash %}
+dnf clean metadata
+{% endhighlight %}
 
 Next upgrade all Foreman packages:
 
-    apt-get update
-    apt-get --only-upgrade install ruby\* foreman\*
-
-You can skip to Step 3.
-
-##### Step 2 (C) - Downloaded release (tar.bz2)
-
-- Uncompress the new program archive in a *new directory*.
-- Copy your database settings-file `config/database.yml` into the new `config` directory.
-
-VERY IMPORTANT: do NOT overwrite `config/settings.yml` with the old one.
-
-##### Step 2 (D) - git checkouts only
+{% highlight bash %}
+dnf upgrade ruby\* foreman\*
+{% endhighlight %}
 
-*Please note* that there can be significant differences between stable and
-development branches.  You MUST take care to select a branch and make sure you
-get the correct one.
+Foreman runs best on PostgreSQL 12 but the default version on EL8 is 10. The installer will try to use version 12 but can't upgrade existing databases. An upgrade can be performed by [switching module streams](https://docs.fedoraproject.org/en-US/modularity/using-modules-switching-streams/).
 
-1. Go to the Foreman root directory and run the following command:
+First make sure you have version 10 installed:
 
-For staying on the stable branch:
+{% highlight bash %}
+rpm -qv postgresql-server
+{% endhighlight %}
 
-- git checkout {{page.version}}-stable
-- git pull
+If you're already on 12 or don't have the PostgreSQL server installed, you can continue with step 3. Otherwise upgrade:
 
-There may have been changes to the ruby gem or node module versions, to update
-run the following commands:
+{% highlight bash %}
+systemctl stop postgresql.service
+dnf module reset postgresql
+dnf module install postgresql:12
+dnf install postgresql-upgrade
+postgresql-setup --upgrade
```

Comment:
Following these instructions I get the same issues I saw previously:

```
[root@pipe-katello-server-nightly-centos8 bats]# postgresql-setup --upgrade
 * Upgrading database.
ERROR: pg_upgrade tool failed
ERROR: Upgrade failed.
 * See /var/lib/pgsql/upgrade_postgresql.log for details.
```

```
[root@pipe-katello-server-nightly-centos8 bats]# cat /var/lib/pgsql/pg_upgrade_server.log 

-----------------------------------------------------------------
  pg_upgrade run on Thu Mar  4 12:52:39 2021
-----------------------------------------------------------------

command: "/usr/lib64/pgsql/postgresql-10/bin/pg_ctl" -w -l "pg_upgrade_server.log" -D "/var/lib/pgsql/data-old" -o "-p 5432 -b  -c listen_addresses='' -c unix_socket_permissions=0700 -c unix_socket_directories='/var/lib/pgsql'" start >> "pg_upgrade_server.log" 2>&1
waiting for server to start....2021-03-04 12:52:39 UTC FATAL:  database files are incompatible with server
2021-03-04 12:52:39 UTC DETAIL:  The data directory was initialized by PostgreSQL version 12, which is not compatible with this version 10.15.
 stopped waiting
pg_ctl: could not start server
Examine the log output.
```

Using:
```
[root@pipe-katello-server-nightly-centos8 bats]# rpm  -q postgresql-upgrade
postgresql-upgrade-12.5-1.module_el8.3.0+620+f1abd701.x86_64
```

That is why I found I had to do the manual move and sed per https://github.com/theforeman/foreman-installer/pull/646#issue-569529894

Reply:
That's very odd. It worked for me. Perhaps I didn't stop the service in my upgrade and that's why it worked? I'll try it again.

Reply:
It works for me with on an empty system:
```console
# dnf install -y postgresql-server
Last metadata expiration check: 0:02:21 ago on Thu 04 Mar 2021 13:24:12 UTC.
Dependencies resolved.
======================================================================================================================================================================================
 Package                                     Architecture                     Version                                                       Repository                           Size
======================================================================================================================================================================================
Installing:
 postgresql-server                           x86_64                           10.15-1.module_el8.3.0+619+dbc95fbc                           appstream                           5.0 M
Installing dependencies:
 libpq                                       x86_64                           12.4-1.el8_2                                                  appstream                           195 k
 postgresql                                  x86_64                           10.15-1.module_el8.3.0+619+dbc95fbc                           appstream                           1.5 M
Enabling module streams:
 postgresql                                                                   10                                                                                                     

Transaction Summary
======================================================================================================================================================================================
Install  3 Packages

Total download size: 6.7 M
Installed size: 26 M
Downloading Packages:
(1/3): libpq-12.4-1.el8_2.x86_64.rpm                                                                                                                  2.5 MB/s | 195 kB     00:00    
(2/3): postgresql-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64.rpm                                                                                      4.8 MB/s | 1.5 MB     00:00    
(3/3): postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64.rpm                                                                               6.8 MB/s | 5.0 MB     00:00    
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                 8.3 MB/s | 6.7 MB     00:00     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                              1/1 
  Installing       : libpq-12.4-1.el8_2.x86_64                                                                                                                                    1/3 
  Installing       : postgresql-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                        2/3 
  Running scriptlet: postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                 3/3 
  Installing       : postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                 3/3 
  Running scriptlet: postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                 3/3 
  Verifying        : libpq-12.4-1.el8_2.x86_64                                                                                                                                    1/3 
  Verifying        : postgresql-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                        2/3 
  Verifying        : postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                 3/3 

Installed:
  libpq-12.4-1.el8_2.x86_64              postgresql-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64              postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64             

Complete!     
# postgresql-setup --initdb
 * Initializing database in '/var/lib/pgsql/data'
 * Initialized, logs are in /var/lib/pgsql/initdb_postgresql.log
# systemctl start postgresql.service 
# sudo -u postgres createdb mydb
could not change directory to "/root": Permission denied
# cd /
# echo 'CREATE TABLE mytable (id INT); INSERT INTO mytable VALUES (1)' | sudo -u postgres psql mydb
CREATE TABLE
INSERT 0 1
# systemctl stop postgresql.service 
# dnf module reset postgresql
Last metadata expiration check: 0:04:40 ago on Thu 04 Mar 2021 13:24:12 UTC.
Dependencies resolved.
======================================================================================================================================================================================
 Package                                     Architecture                               Version                                     Repository                                   Size
======================================================================================================================================================================================
Resetting modules:
 postgresql                                                                                                                                                                          

Transaction Summary
======================================================================================================================================================================================

Is this ok [y/N]: y
Complete!
# dnf module install postgresql:12
Last metadata expiration check: 0:04:46 ago on Thu 04 Mar 2021 13:24:12 UTC.
Dependencies resolved.
======================================================================================================================================================================================
 Package                                     Architecture                     Version                                                       Repository                           Size
======================================================================================================================================================================================
Upgrading:
 postgresql                                  x86_64                           12.5-1.module_el8.3.0+620+f1abd701                            appstream                           1.5 M
 postgresql-server                           x86_64                           12.5-1.module_el8.3.0+620+f1abd701                            appstream                           5.6 M
Installing module profiles:
 postgresql/server                                                                                                                                                                   
Enabling module streams:
 postgresql                                                                   12                                                                                                     

Transaction Summary
======================================================================================================================================================================================
Upgrade  2 Packages

Total download size: 7.1 M
Is this ok [y/N]: y
Downloading Packages:
(1/2): postgresql-12.5-1.module_el8.3.0+620+f1abd701.x86_64.rpm                                                                                       5.0 MB/s | 1.5 MB     00:00    
(2/2): postgresql-server-12.5-1.module_el8.3.0+620+f1abd701.x86_64.rpm                                                                                7.4 MB/s | 5.6 MB     00:00    
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                 8.6 MB/s | 7.1 MB     00:00     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                              1/1 
  Running scriptlet: postgresql-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                         1/1 
  Upgrading        : postgresql-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                         1/4 
  Running scriptlet: postgresql-server-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                  2/4 
  Upgrading        : postgresql-server-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                  2/4 
  Running scriptlet: postgresql-server-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                  2/4 
  Running scriptlet: postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                 3/4 
  Cleanup          : postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                 3/4 
  Running scriptlet: postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                 3/4 
  Cleanup          : postgresql-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                        4/4 
  Running scriptlet: postgresql-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                        4/4 
  Verifying        : postgresql-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                         1/4 
  Verifying        : postgresql-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                        2/4 
  Verifying        : postgresql-server-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                  3/4 
  Verifying        : postgresql-server-10.15-1.module_el8.3.0+619+dbc95fbc.x86_64                                                                                                 4/4 

Upgraded:
  postgresql-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                   postgresql-server-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                  

Complete!
# dnf install postgresql-upgrade
Last metadata expiration check: 0:04:54 ago on Thu 04 Mar 2021 13:24:12 UTC.
Dependencies resolved.
======================================================================================================================================================================================
 Package                                      Architecture                     Version                                                      Repository                           Size
======================================================================================================================================================================================
Installing:
 postgresql-upgrade                           x86_64                           12.5-1.module_el8.3.0+620+f1abd701                           appstream                           4.1 M
Installing dependencies:
 perl-Carp                                    noarch                           1.42-396.el8                                                 baseos                               30 k
 perl-Exporter                                noarch                           5.72-396.el8                                                 baseos                               34 k
 perl-libs                                    x86_64                           4:5.26.3-416.el8                                             baseos                              1.6 M
 tcl                                          x86_64                           1:8.6.8-2.el8                                                baseos                              1.1 M

Transaction Summary
======================================================================================================================================================================================
Install  5 Packages

Total download size: 6.8 M
Installed size: 22 M
Is this ok [y/N]: y
Downloading Packages:
(1/5): perl-Carp-1.42-396.el8.noarch.rpm                                                                                                              1.1 MB/s |  30 kB     00:00    
(2/5): perl-Exporter-5.72-396.el8.noarch.rpm                                                                                                          1.1 MB/s |  34 kB     00:00    
(3/5): tcl-8.6.8-2.el8.x86_64.rpm                                                                                                                     3.5 MB/s | 1.1 MB     00:00    
(4/5): perl-libs-5.26.3-416.el8.x86_64.rpm                                                                                                            3.6 MB/s | 1.6 MB     00:00    
(5/5): postgresql-upgrade-12.5-1.module_el8.3.0+620+f1abd701.x86_64.rpm                                                                               5.2 MB/s | 4.1 MB     00:00    
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                 3.7 MB/s | 6.8 MB     00:01     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                              1/1 
  Installing       : perl-Exporter-5.72-396.el8.noarch                                                                                                                            1/5 
  Installing       : perl-libs-4:5.26.3-416.el8.x86_64                                                                                                                            2/5 
  Installing       : perl-Carp-1.42-396.el8.noarch                                                                                                                                3/5 
  Installing       : tcl-1:8.6.8-2.el8.x86_64                                                                                                                                     4/5 
  Running scriptlet: tcl-1:8.6.8-2.el8.x86_64                                                                                                                                     4/5 
  Installing       : postgresql-upgrade-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                 5/5 
  Running scriptlet: postgresql-upgrade-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                 5/5 
  Verifying        : postgresql-upgrade-12.5-1.module_el8.3.0+620+f1abd701.x86_64                                                                                                 1/5 
  Verifying        : perl-Carp-1.42-396.el8.noarch                                                                                                                                2/5 
  Verifying        : perl-Exporter-5.72-396.el8.noarch                                                                                                                            3/5 
  Verifying        : perl-libs-4:5.26.3-416.el8.x86_64                                                                                                                            4/5 
  Verifying        : tcl-1:8.6.8-2.el8.x86_64                                                                                                                                     5/5 

Installed:
  perl-Carp-1.42-396.el8.noarch      perl-Exporter-5.72-396.el8.noarch      perl-libs-4:5.26.3-416.el8.x86_64      postgresql-upgrade-12.5-1.module_el8.3.0+620+f1abd701.x86_64     
  tcl-1:8.6.8-2.el8.x86_64          

Complete!
# postgresql-setup --upgrade
 * Upgrading database.
 * Upgraded OK.
WARNING: The configuration files were replaced by default configuration.
WARNING: The previous configuration and data are stored in folder
WARNING: /var/lib/pgsql/data-old.
WARNING: If you've just upgraded your database from a previous major version of
Fedora or RHEL, please run reindexdb against your databases.  Core library
collation data may have changed and this will invalidate database indexes.  For
example, in Fedora 28 and RHEL 8 there have been extensive changes in glibc
collations to support ISO 14651:2016 (Unicode 9.0.0 data) and your indexes may
be affected: https://sourceware.org/ml/libc-announce/2018/msg00002.html
 * See /var/lib/pgsql/upgrade_postgresql.log for details.
# systemctl status postgresql.service 
● postgresql.service - PostgreSQL database server
   Loaded: loaded (/usr/lib/systemd/system/postgresql.service; disabled; vendor preset: disabled)
   Active: inactive (dead)

Mar 04 13:27:24 centos8.wisse.example.com postmaster[23604]: 2021-03-04 13:27:24.457 UTC [23604] LOG:  listening on IPv4 address "127.0.0.1", port 5432
Mar 04 13:27:24 centos8.wisse.example.com postmaster[23604]: 2021-03-04 13:27:24.461 UTC [23604] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
Mar 04 13:27:24 centos8.wisse.example.com postmaster[23604]: 2021-03-04 13:27:24.467 UTC [23604] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
Mar 04 13:27:24 centos8.wisse.example.com postmaster[23604]: 2021-03-04 13:27:24.486 UTC [23604] LOG:  redirecting log output to logging collector process
Mar 04 13:27:24 centos8.wisse.example.com postmaster[23604]: 2021-03-04 13:27:24.486 UTC [23604] HINT:  Future log output will appear in directory "log".
Mar 04 13:27:24 centos8.wisse.example.com systemd[1]: Started PostgreSQL database server.
Mar 04 13:28:29 centos8.wisse.example.com systemd[1]: Stopping PostgreSQL database server...
Mar 04 13:28:29 centos8.wisse.example.com systemd[1]: postgresql.service: Killing process 23605 (postmaster) with signal SIGKILL.
Mar 04 13:28:29 centos8.wisse.example.com systemd[1]: postgresql.service: Succeeded.
Mar 04 13:28:29 centos8.wisse.example.com systemd[1]: Stopped PostgreSQL database server.
# systemctl start postgresql.service 
# echo 'SELECT * FROM mytable' | sudo -u postgres psql mydb
 id 
----
  1
(1 row)

```

Reply:
Correct. The problem comes from our puppet modules. The puppet-postgresql module sets the datadir (https://github.com/puppetlabs/puppetlabs-postgresql/blob/849832666c0352d59c93d2fdbcde8c4bcd401923/manifests/params.pp#L57) in `postgresql.conf` as a hard-coded value. So when the `postgresql-upgrade` tool moves `/var/lib/pgsql/data` to `/var/lib/pgsql/data-old` it retains the `postgresql.conf` that is pointing at `/var/lib/pgsql/data` as the datadir. Then when the upgrade tooling attempts to read the `version` of the old data directory (i.e. `/var/lib/pgsql/data-old`) it is pointed at `/var/lib/pgsql/data` by `postgresql.conf` and retrieves an incorrect version. That is why this step is neccessary in my procedure:

```
sed -i "s/\/var\/lib\/pgsql\/data/\/var\/lib\/pgsql\/data-old/g" data-old/postgresql.conf
```

And why manual use of `pg_upgrade` is needed since `postgresql-upgrade` is all in one with assumptions of structure.

Reply:
I opened https://bugzilla.redhat.com/show_bug.cgi?id=1935301 for this. How about the additional commit I pushed?

Reply:
That seems to work -- I updated the automation in FOC to mimic this procedure (https://github.com/theforeman/foreman-operations-collection/pull/21/files#diff-f66848610d4b546cc152eeb0ff52d94bfc4928543e78cc66bfcb9e733f7ba690R10). Note that I had to do something similar to when we did EL7 postgresql upgrades and set some collate and locale properties which I think are just due to the runtime environment. You'll have to take my word that this passes locally because right now CI is having fits due to unrelated things.